packages/jaeger-ui/src/model/ddg/types.tsx

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+export type TDdgPayload = {
+  operation: string;
+  service: string;
+}[][];
+
+export type TDdgService = {
+  name: string;
+  operations: Map<string, TDdgOperation>;
+};
+
+export type TDdgOperation = {
+  name: string;
+  pathElems: PathElem[];
+  service: TDdgService;
+};
+
+export type TDdgPath = {
+  focalIdx: number;
+  members: PathElem[];
+};
+
+export type TDdgServiceMap = Map<string, TDdgService>;
+
+export type TDdgPathElemsByDistance = Map<number, PathElem[]>;
+
+export type TDdgTransformedDdgData = {
+  pathElemsByDistance: TDdgPathElemsByDistance;
+  paths: TDdgPath[];
+  services: TDdgServiceMap;
+};
+
+export class PathElem {
+  memberOf: TDdgPath;
+  operation: TDdgOperation;
+  pathIdx: number;
+  private _visibilityIdx?: number;
+
+  constructor({ path, operation, pathIdx }: { path: TDdgPath; operation: TDdgOperation; pathIdx: number }) {
+    this.memberOf = path;
+    this.operation = operation;
+    this.pathIdx = pathIdx;
+
+    operation.pathElems.push(this);
+  }
+
+  get distance() {
+    return this.pathIdx - this.memberOf.focalIdx;
```

Comment:
I like this, but on second thought, the `PathElem` knowing details about `TDdgPath` like `focalIdx`, and being dependent on how focal the index is represented, is unfortunate.

Reply:
I was thinking of making `TDdgPath` a class as well, so that `focalIdx` can be a getter. The benefit of that would be the ability to enable/disable distinguishing operations without needing to recalculate all `focalIdx`s.
It is somewhat coupled, but the value is high. Through TypeScript (and good tests) it's pretty clear if/when this coupling is broken.

Reply:
Scratch what I said about `focalIdx` as a getter, since if the `focalIdx` changes as a result of the `focalNode` changing, all `visibilityIdx`s would be invalid.